File: android/src/main/java/app/capgo/networkdiagnostics/NetworkDiagnosticsPlugin.java
package app.capgo.networkdiagnostics;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@CapacitorPlugin(name = "NetworkDiagnostics")
public class NetworkDiagnosticsPlugin extends Plugin {

    private final NetworkDiagnostics implementation = new NetworkDiagnostics();
    private final ExecutorService executor = Executors.newCachedThreadPool();

    @PluginMethod
    public void getNetworkStatus(PluginCall call) {
        executor.execute(() -> call.resolve(implementation.getNetworkStatus(getContext())));
    }

    @PluginMethod
    public void testUrl(PluginCall call) {
        String url = call.getString("url");
        if (url == null || url.isEmpty()) {
            call.reject("URL is required");
            return;
        }

        String method = call.getString("method", "HEAD");
        Integer timeoutMs = call.getInt("timeoutMs", 10000);
        Boolean followRedirects = call.getBoolean("followRedirects", true);

        executor.execute(() ->
            call.resolve(
                implementation.testUrl(url, method, timeoutMs == null ? 10000 : timeoutMs, followRedirects == null || followRedirects)
            )
        );
    }

    @PluginMethod
    public void testPort(PluginCall call) {
        String host = call.getString("host");
        Integer port = call.getInt("port");
        if (host == null || host.isEmpty()) {
            call.reject("Host is required");
            return;
        }
        if (port == null || port <= 0 || port > 65535) {
            call.reject("Port must be between 1 and 65535");
            return;
        }

        Integer timeoutMs = call.getInt("timeoutMs", 5000);
        executor.execute(() -> call.resolve(implementation.testPort(host, port, timeoutMs == null ? 5000 : timeoutMs)));
    }

    @PluginMethod
    public void testWebSocket(PluginCall call) {
        String url = call.getString("url");
        if (url == null || url.isEmpty()) {
            call.reject("URL is required");
            return;
        }

        Integer timeoutMs = call.getInt("timeoutMs", 10000);
        executor.execute(() -> call.resolve(implementation.testWebSocket(url, timeoutMs == null ? 10000 : timeoutMs)));
    }

    @PluginMethod
    public void testDownloadSpeed(PluginCall call) {
        String url = call.getString("url");
        if (url == null || url.isEmpty()) {
            call.reject("URL is required");
            return;
        }

        Integer maxBytes = call.getInt("maxBytes", 5 * 1024 * 1024);
        Integer timeoutMs = call.getInt("timeoutMs", 30000);
        executor.execute(() ->
            call.resolve(
                implementation.testDownloadSpeed(url, maxBytes == null ? 5 * 1024 * 1024 : maxBytes, timeoutMs == null ? 30000 : timeoutMs)
            )
        );
    }

    @PluginMethod
    public void testPacketLoss(PluginCall call) {
        String mode = call.getString("mode", "");
        String host = call.getString("host", "");
        String url = call.getString("url", "");
        Integer port = call.getInt("port", 0);
        Integer count = call.getInt("count", 10);
        Integer timeoutMs = call.getInt("timeoutMs", 3000);
        Integer intervalMs = call.getInt("intervalMs", 250);

        if ((host == null || host.isEmpty() || port == null || port <= 0) && (url == null || url.isEmpty())) {
            call.reject("Either host and port, or url is required");
            return;
        }

        executor.execute(() ->
            call.resolve(
                implementation.testPacketLoss(
                    mode == null ? "" : mode,
                    host == null ? "" : host,
                    port == null ? 0 : port,
                    url == null ? "" : url,
                    count == null ? 10 : count,
                    timeoutMs == null ? 3000 : timeoutMs,
                    intervalMs == null ? 250 : intervalMs
                )
            )
        );
    }

    @PluginMethod
    public void runDiagnostics(PluginCall call) {
        JSArray urls = call.getArray("urls", new JSArray());
        JSArray ports = call.getArray("ports", new JSArray());
        JSArray websockets = call.getArray("websockets", new JSArray());
        JSObject download = call.getObject("download");
        JSObject packetLoss = call.getObject("packetLoss");

        executor.execute(() -> call.resolve(implementation.runDiagnostics(getContext(), urls, ports, websockets, download, packetLoss)));
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("version", implementation.getPluginVersion());
        call.resolve(ret);
    }

    @Override
    protected void handleOnDestroy() {
        super.handleOnDestroy();
        executor.shutdownNow();
    }
}

File: ios/Sources/NetworkDiagnosticsPlugin/NetworkDiagnosticsPlugin.swift
import Foundation
import Capacitor

@objc(NetworkDiagnosticsPlugin)
public class NetworkDiagnosticsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NetworkDiagnosticsPlugin"
    public let jsName = "NetworkDiagnostics"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getNetworkStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testPort", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testWebSocket", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testDownloadSpeed", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testPacketLoss", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "runDiagnostics", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = NetworkDiagnostics()

    @objc func getNetworkStatus(_ call: CAPPluginCall) {
        Task {
            call.resolve(await implementation.getNetworkStatus())
        }
    }

    @objc func testUrl(_ call: CAPPluginCall) {
        guard let url = call.getString("url"), !url.isEmpty else {
            call.reject("URL is required")
            return
        }

        Task {
            call.resolve(await implementation.testUrl(
                url: url,
                method: call.getString("method") ?? "HEAD",
                timeoutMs: call.getInt("timeoutMs") ?? 10_000,
                followRedirects: call.getBool("followRedirects") ?? true
            ))
        }
    }

    @objc func testPort(_ call: CAPPluginCall) {
        guard let host = call.getString("host"), !host.isEmpty else {
            call.reject("Host is required")
            return
        }
        let port = call.getInt("port") ?? 0
        guard port > 0 && port <= 65_535 else {
            call.reject("Port must be between 1 and 65535")
            return
        }

        Task {
            call.resolve(await implementation.testPort(
                host: host,
                port: port,
                timeoutMs: call.getInt("timeoutMs") ?? 5_000
            ))
        }
    }

    @objc func testWebSocket(_ call: CAPPluginCall) {
        guard let url = call.getString("url"), !url.isEmpty else {
            call.reject("URL is required")
            return
        }

        Task {
            call.resolve(await implementation.testWebSocket(
                url: url,
                timeoutMs: call.getInt("timeoutMs") ?? 10_000
            ))
        }
    }

    @objc func testDownloadSpeed(_ call: CAPPluginCall) {
        guard let url = call.getString("url"), !url.isEmpty else {
            call.reject("URL is required")
            return
        }

        Task {
            call.resolve(await implementation.testDownloadSpeed(
                url: url,
                maxBytes: call.getInt("maxBytes") ?? 5 * 1024 * 1024,
                timeoutMs: call.getInt("timeoutMs") ?? 30_000
            ))
        }
    }

    @objc func testPacketLoss(_ call: CAPPluginCall) {
        let host = call.getString("host") ?? ""
        let port = call.getInt("port") ?? 0
        let url = call.getString("url") ?? ""

        guard (!host.isEmpty && port > 0) || !url.isEmpty else {
            call.reject("Either host and port, or url is required")
            return
        }

        Task {
            let options = PacketLossOptions(
                mode: call.getString("mode") ?? "",
                host: host,
                port: port,
                url: url,
                count: call.getInt("count") ?? 10,
                timeoutMs: call.getInt("timeoutMs") ?? 3_000,
                intervalMs: call.getInt("intervalMs") ?? 250
            )
            call.resolve(await implementation.testPacketLoss(options))
        }
    }

    @objc func runDiagnostics(_ call: CAPPluginCall) {
        let urls = call.getArray("urls", []) as? [[String: Any]] ?? []
        let ports = call.getArray("ports", []) as? [[String: Any]] ?? []
        let websockets = call.getArray("websockets", []) as? [[String: Any]] ?? []
        let download = call.getObject("download")
        let packetLoss = call.getObject("packetLoss")

        Task {
            call.resolve(await implementation.runDiagnostics(
                urls: urls,
                ports: ports,
                websockets: websockets,
                download: download,
                packetLoss: packetLoss
            ))
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }
}
